ActiveDocuments/EditContext/dev-design.md

```diff
@@ -0,0 +1,105 @@
+2. These are delivered to the browser, which opted-in to using the system's text services framework in order to integrate with the IMEs installed on the system. This will cause input to be forwarded to the active IME. 
+3. The IME is then able to query the text services to read contextual information related to the underlying editable text in order to provide suggestions, and potentially modify which character(s) should be written to the shared buffer. 
+4. These modifications are typically performed based on the current selection, which is also communicated through the text services framework. 
+5. When the shared buffer is updated, the web application will be notified of this via the ```textupdate``` event.
+
+
+This section describes the sequences of events that get fired on the EditContext and focused element when EditContext has focus and IME is active. In this event sequence, the user types in two characters, then commits to the first IME candidate by hitting 'Space'.
+
+|  Event                | EventTarget        |  Related key in sequence
+| -------------         | -----------------  | -------------------
+|  keydown              | focused element    |  Key 1
+|  compositionstart     | active EditContext |  ...
+|  textupdate           | active EditContext |  ...
+|  textformatupdate     | active EditContext |  ...
+|  keyup                | focused element    |  ...
+|  keydown              | focused element    |  Key 2
+|  textupdate           | active EditContext |  ...
+|  textformatupdate     | active EditContext |  ...
+|  keyup                | focused element    |  ...
+|  keydown              | focused element    |  Space
+|  textupdate           | active EditContext |  (committed IME characters available in event.updateText)
+|  keyup                | focused element    |  ...
+|  compositionend       | active EditContext |
+
+Note that the composition events are also not fired on the focused element as the composition is operating on the shared buffer that is represented by the EditContext.
+
+### Externally triggered changes
+
+Changes to the editable contents can also come from external events, such as collaboration scenarios. In this case, the web editing framework may get some XHR completion that notifies it of some pending collaboartive change that another user has committed. The framework is then responsible for writing to the shared buffer, via the ```updateText()``` method.
+
+![external input](external_input.png)
+
+## API Details
+
+```focus``` is used to activate an EditContext and a way to tell the OS that the author wants advanced text input methods enabled and that the metadata required for suggestions, where text input UI should appear, what the input mode is, etc. should all come from this particular editContext instance. This will create a strong reference (internally) from document to that editContext. The typical owner of an editContext instance will be the web app or widget that needs to enable text input.
```

Comment:
Nit on back-tick usage: 

```suggestion
`focus` is used to activate an EditContext and a way to tell the OS that the author wants advanced text input methods enabled and that the metadata required for suggestions, where text input UI should appear, what the input mode is, etc. should all come from this particular editContext instance. This will create a strong reference (internally) from document to that editContext. The typical owner of an editContext instance will be the web app or widget that needs to enable text input.
```